tpm2_pytss/types.py

```diff
@@ -921,54 +921,55 @@ def Unmarshal(cls, buf):
 
 
 class TPM2B_OBJECT(TPM_OBJECT):
+    @classmethod
+    def _get_bytefield(cls):
+        tipe = ffi.typeof(f"{cls.__name__}")
+        for f in tipe.fields:
+            if f[0] != "size":
+                return f[0]
+
     def __init__(self, _cdata=None, **kwargs):
         _cdata, kwargs = fixup_cdata_kwargs(self, _cdata, kwargs)
         super().__init__(_cdata, **kwargs)
 
-        # we don't set size on nested compount TPM2B structures, like TPM2B_PUBLIC
-        tipe = ffi.typeof(_cdata)
-        if "TPM2B_DIGEST" not in tipe.cname:
-            return
-
-        # if their is nothing to update OR the caller passed both the
-        # buffer AND the size, their is nothing to do.
-        if len(kwargs) == 0 or len(kwargs) == 2:
-            return
-
-        key = [*kwargs][0]
-        value = kwargs[key]
-        if len(kwargs) == 0:
-            return
-
-        # update the len
-        _cdata.size = len(value)
+    def __setattr__(self, key, value):
+        _bytefield = type(self)._get_bytefield()
+        if key == "size":
+            raise AttributeError(f"{key} is read only")
+        super().__setattr__(key, value)
+        if key == _bytefield:
+            self._cdata.size = len(value)
 
     def __getattribute__(self, key):
-        value = super().__getattribute__(key)
-        if key == "buffer":
-            value = ffi.buffer(value, self._cdata.size)
-        return value
+        _bytefield = type(self)._get_bytefield()
+        if key == _bytefield:
+            b = getattr(self._cdata, _bytefield)
+            return memoryview(ffi.buffer(b, self._cdata.size))
+        return super().__getattribute__(key)
 
     def __len__(self):
-        cdata = self._cdata
-        return cdata.size
+        return self._cdata.size
+
+    def __getitem__(self, index):
+        _bytefield = type(self)._get_bytefield()
+        buf = getattr(self, _bytefield)
+        if isinstance(index, int):
+            if index >= self._cdata.size:
+                raise IndexError("out of range")
+            return buf[index]
+        elif isinstance(index, slice):
+            return buf[index]
+        else:
+            raise TypeError("index must an int or a slice")
 
     def __bytes__(self):
-
-        _cdata = self._cdata
-
-        # not everything uses .buffer, for instance TPM2B_NAME uses .name
-        tipe = cpointer_to_ctype(_cdata)
-        field_name = next((v[0] for v in tipe.fields if v[0] != "size"), None)
-
-        buffer = TPM2B_unpack(_cdata, n=field_name)
```

Comment:
TPM2B_unpack is no longer used remove from imports

Reply:
I need to touch ESAPI.py as well when removing TPM2B_unpack, so will make another commit (under this PR) for that